lib/ddtrace/contrib/redis/configuration/resolver.rb

```diff
@@ -0,0 +1,38 @@
+module Datadog
+  module Contrib
+    module Redis
+      module Configuration
+        # Converts Symbols, Strings, and Hashes to a normalized connection settings Hash.
+        class Resolver < Contrib::Configuration::Resolver
+          def resolve(key_or_hash)
+            return :default if key_or_hash == :default
+
+            normalize(key_or_hash)
+          end
+
+          def normalize(hash)
+            resolved_configuration = resolve_configuration(hash)
+            return { url: resolved_configuration[:url] } if resolved_configuration[:scheme] == 'unix'
+
+            # Connexion strings are always converted to host, port, db and scheme
+            # but the host, port, db and scheme will generate the :url only after
+            # establishing a first connexion
+            {
+              host: resolved_configuration[:host],
+              port: resolved_configuration[:port],
+              db: resolved_configuration[:db],
+              scheme: resolved_configuration[:scheme]
+            }
+          end
+
+          # The option parsing in Redis::Client is implemented as a instance method
+          # of the client itself. Since it cannot be imported from a library module
+          # the configuration will be resolved within a new redis instance
+          def resolve_configuration(options)
+            ::Redis::Client.new(options).options
```

Comment:
Although it keeps things very practical and ensures consistency, I'm worried this could have some bad side effects (e.g. connections automatically made, etc.) It would be best to avoid instantiating a client for some alternative, even if it means copying code.

I'm thinking maybe there's a way to avoid this issue though... if we think about it from the perspective of "access" use, we already have `options` from the client instance itself, so we can just pass that in and `normalize` it.

I think the bigger question is the "configure" use case, where the user is providing some hash that isn't a proper `Redis::Client` options hash; some of the fields might be incomplete, missing, or unexpanded. Since `url`, `host`, `port`, `db`, and `scheme` are the only fields that matter, are any of these mutated at all by passing them through `Redis::Client.new`? e.g. does `localhost` become `127.0.0.1`? If none of them go under mutation, then all we'd need to do is make sure the Redis defaults sub in place for the missing fields.

I'm going to think about how to express this in code better and get back to this.

Reply:
> Although it keeps things very practical and ensures consistency, I'm worried this could have some bad side effects (e.g. connections automatically made, etc.) It would be best to avoid instantiating a client for some alternative, even if it means copying code.

A `Redis::Client` instantiation does not open any connection. It preprocess the options using the `_parse_options` instance method and it has been almost unchanged since 2012.
The function can be copied here, if it is the standard way you did for other integrations

Reply:
> I'm thinking maybe there's a way to avoid this issue though... if we think about it from the perspective of "access" use, we already have options from the client instance itself, so we can just pass that in and normalize it.

When the resolver resolves the configuration passed to the `describes:` key, we do not have yet any redis client instance. We may backport the parse options functions with the defaults and make sure we can get the same result without instantiating any redis client

Reply:
> I think the bigger question is the "configure" use case, where the user is providing some hash that isn't a proper Redis::Client options hash; some of the fields might be incomplete, missing, or unexpanded. Since url, host, port, db, and scheme are the only fields that matter, are any of these mutated at all by passing them through Redis::Client.new? e.g. does localhost become 127.0.0.1? If none of them go under mutation, then all we'd need to do is make sure the Redis defaults sub in place for the missing fields.

This was illustrated in commit 70ed4fc .

The option hash passed to the instance is enriched with the defaults.

If the `url` is passed, it would be used for the connection and for generating an internal options hash with db, host etc (unless the url scheme is Unix).
If the client is initialized with an option hash, the url will be generated at the first connexion so it will be nil until then.
The third way to set the url is via env var.

In the configuration hash localhost never becomes 127.0.0.1. 

The match is then safe if:
- we use only the url for unix scheme url
- we use a subset of the configuration hash in all other cases